_posts/2022-06-17-nativejdb-debugger-for-native-images.adoc

```diff
@@ -0,0 +1,143 @@
+---
+layout: post
+title: 'Debugging natively compiled Java code with NativeJDB'
+date: 2022-06-22
+tags: java debugger native cloud container 
+synopsis: NativeJDB bridges the gap between the Java Debugger (JDB) framework available in modern IDEs and native debugging via GNU Project debugger (GDB).
+author: aavarghese, vazirim
+---
+:imagesdir: /assets/images/posts/nativejdb
+
+image::nativejdb.png[alt=NativeJDB]
+
+==== In collaboration with:
+
+
+_Max Andersen, Dimitris Andreadis, Andrew Dinn, Stuart Douglas, Jason Greene, David Grove, David Lloyd, Thomas Qvarnstrom, Foivos Zakkak, Galder Zamarreno (IBM Research and Red Hat)_
+
+
+{empty} +
+
+
+https://quarkus.io/[Quarkus] is a cloud-native Java development framework, which allows Java code to be mapped to Kubernetes containers and compiled natively. https://www.graalvm.org/22.1/reference-manual/native-image/[Native compilation] is useful for serverless computing, it avoids the overhead of running a JVM in containers and lets us execute serverless code directly. Debugging natively compiled code is not an easy task. It can be very different from the original Java source code due to extensive code optimizations, as well as differences in libraries used in native vs non-natively compiled code. In some cases, it is necessary to view both the source and the assembly code in order to gain better understanding of the program, further complicating the debugging task. 
+
+
+https://developers.redhat.com/blog/2020/06/25/debugging-graalvm-native-images-using-gdb[Native executables can be debugged with GDB], which is a C/C++ debugger in combination with a tool like emacs to step through the source code. However, these tools may not be familiar to a Java developer. Recently, extensions to https://marketplace.visualstudio.com/items?itemName=oracle-labs-graalvm.graalvm[VSCode] and https://plugins.jetbrains.com/plugin/19237-graalvm-native-debugger[IntelliJ] have been developed to alleviate this problem. Both these tools are specific to the IDE they extend, and need to be used in a Linux environment due to current limitations of Java native compilers. 
+
+
+https://github.com/nativejdb/[NativeJDB] is an *open-source* tool that supports *remote debugging* of a natively compiled Java code. In principle, this allows debugging from any IDE that supports the Java Platform Debugger Architecture (JPDA), such as IntelliJ, Eclipse or VSCode. Attaching remotely means that the user can start a debugging session on *any O/S*, while the NativeJDB process itself runs in a Linux container. Our challenge is to bridge the gap between JPDA and the GDB, two stacks that don't typically talk to each other! To achieve this we essentially teach the GDB to speak the Java Debug Wire Protocol (JDWP), used by IDEs to communicate with Java debuggers.
```

Comment:
does this mean it only works on debugging linux binaries? no osx nor windows debug hosts ? just want to clarify if that is the implication.

Reply:
yes, native compilation with graalvm works only on Linux at the moment.